src/braintree-manager/payment-providers.ts

```diff
@@ -0,0 +1,136 @@
+import { PaymentClientsInterface } from './payment-clients';
+import { GooglePayHandlerInterface, GooglePayHandler } from './payment-providers/google-pay';
+import { BraintreeManagerInterface, HostingEnvironment } from './braintree-interfaces';
+import { HostedFieldConfiguration } from './payment-providers/credit-card/hosted-field-configuration';
+
+export interface PaymentProvidersInterface {
+  creditCardHandler: PromisedSingleton<CreditCardHandlerInterface>;
+  applePayHandler: PromisedSingleton<ApplePayHandlerInterface>;
+  venmoHandler: PromisedSingleton<VenmoHandlerInterface | undefined>;
+  paypalHandler: PromisedSingleton<PayPalHandlerInterface>;
+  googlePayHandler: PromisedSingleton<GooglePayHandlerInterface>;
+}
+
+/**
+ * The PaymentProviders class contains the IA-specific handlers for each of the
+ * different payment providers.
+ *
+ * They are generally data-focused, as opposed to UI-focused, but there
+ * is some cross-polination between PaymentProviders and FlowHandlers.
+ *
+ * @export
+ * @class PaymentProviders
+ * @implements {PaymentProvidersInterface}
+ */
+export class PaymentProviders implements PaymentProvidersInterface {
+  creditCardHandler: PromisedSingleton<CreditCardHandlerInterface>;
+
+  applePayHandler: PromisedSingleton<ApplePayHandlerInterface>;
+
+  venmoHandler: PromisedSingleton<VenmoHandlerInterface | undefined>;
+
+  paypalHandler: PromisedSingleton<PayPalHandlerInterface>;
+
+  googlePayHandler: PromisedSingleton<GooglePayHandlerInterface>;
+
+  private braintreeManager: BraintreeManagerInterface;
+
+  private venmoProfileId?: string;
+
+  private googlePayMerchantId?: string;
+
+  private hostedFieldConfig: HostedFieldConfiguration;
+
+  private hostingEnvironment: HostingEnvironment = HostingEnvironment.Development;
+
+  private paymentClients: PaymentClientsInterface;
+
+  constructor(options: {
+    braintreeManager: BraintreeManagerInterface;
+    paymentClients: PaymentClientsInterface;
+    venmoProfileId?: string;
+    googlePayMerchantId?: string;
+    hostingEnvironment: HostingEnvironment;
+    hostedFieldConfig: HostedFieldConfiguration;
+  }) {
+    this.braintreeManager = options.braintreeManager;
+    this.venmoProfileId = options.venmoProfileId;
+    this.googlePayMerchantId = options.googlePayMerchantId;
+    this.paymentClients = options.paymentClients;
+    this.hostingEnvironment = options.hostingEnvironment;
+    this.hostedFieldConfig = options.hostedFieldConfig;
+
+    this.creditCardHandler = new PromisedSingleton<CreditCardHandlerInterface>({
+      generator: this.paymentClients.hostedFields.get().then(client => {
+        return new CreditCardHandler({
+          braintreeManager: this.braintreeManager,
+          hostedFieldClient: client,
+          hostedFieldConfig: this.hostedFieldConfig,
+        });
+      }),
+    });
+
+    this.applePayHandler = new PromisedSingleton<ApplePayHandlerInterface>({
+      generator: this.paymentClients.applePay.get().then(client => {
```

Comment:
what is the purpose of using `generator`? will we be iterating over something?

Reply:
The `generator` in the `PromisedSingleton` is the thing that "generates" the singleton. So like we're passing in a Promise that will resolve to some object (in this case an `ApplePayHandlerInterface`).

Anything that calls `applePayHandler.get()` will always get the same `ApplePayHandlerInterface`